src/transformers/vocabulary2id.py

```diff
@@ -0,0 +1,71 @@
+import operator
+
+from pyspark import RDD, Row
+from models.code2vec import Code2Vec
+
+from ast import literal_eval as make_tuple
+from sourced.ml.transformers import Transformer
+
+
+class Vocabulary2Id(Transformer):
+    def __init__(self, keep_freq: bool, output: str, **kwargs):
+        super().__init__(**kwargs)
+        self.keep_freq = keep_freq
+        self.output = output
+
+    def __call__(self, rows: RDD):
+        values, paths = self.build_vocabularies(rows)
+
+        doc2path_contexts = self.build_doc2pc(rows)
+
+        values = values.collect()
+        paths = paths.collect()
+
+        doc2path_contexts = doc2path_contexts.collect()
+
+        Code2Vec().construct(values=values, paths=paths, path_contexts=doc2path_contexts).save(
+            self.output)
+
+    @staticmethod
+    def _unstringify_path_context(row):
+        """
+        Takes a row containing ((pc, doc), freq) and returns a tuple (removes namespace prefix v.)
+        """
+        return make_tuple(row[0][0][2:])
+
+    def _get_path(self, row: Row):
+        # 2: removes the namespace v. from the string to parse it as tuple
+        k, v = Vocabulary2Id._unstringify_path_context(row), row[1]
+        return [(k[0], 1), (k[1], 1), (k[2], 1)]
+
+    def build_vocabularies(self, rows: RDD):
+        """
+        Process rows to gather values and paths with/without their frequencies.
+        """
+
+        r = rows \
+            .flatMap(self._get_path) \
+            .reduceByKey(operator.add)
+
+        get_elem = lambda x: x[0]
+        if not self.keep_freq:
+            r = r.map(lambda x: x[0])
+            get_elem = lambda x: x
+
+        values = r.filter(lambda x: type(get_elem(x)) == str)
+        paths = r.filter(lambda x: type(get_elem(x)) == tuple)
+
+        return values, paths
+
+    def _doc2pc(self, row: Row):
+        pc, doc = Vocabulary2Id._unstringify_path_context(row), row[0][1]
+
+        return doc, [pc]
+
+    def build_doc2pc(self, rows: RDD):
+        """
+        Process rows and build elements (doc, [path_context_1, path_context_2, ...])
+        """
+        return rows \
+            .map(self._doc2pc) \
+            .reduceByKey(operator.add)
```

Comment:
if you want to do this, you need to map the RDD to `(doc, [pc])`, 1 no ? In all cases I think for now it is best to not add calculations for document/featurefreqs, so simply do a distinct after the map (distinct should be done after mapping of the path-contrxes strings to indexes, see comment above)

Reply:
the idea was to add the lists for each key to have a `(doc, [pc_1, pc_2, ...])`  because _doc2pc is a single path context lik `(doc, [pc])`

Reply:
okay i see now, problem is this might give you cases like this: `(doc, [pc1, pc2, pc1, pc3 ...])` so you ought to still do a distinct, before the `reduceByKey`